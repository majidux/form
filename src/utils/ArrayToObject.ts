export function handleArrayToObject(data: any[]): any {
  return data.reduce((output: any, current: any) => ({
    ...output,
    ...current,
  }));
}
