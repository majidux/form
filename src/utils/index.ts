import * as yup from 'yup';
export { yup };

export { Children, useState } from 'react';
export { Form } from '../components/Form';
export { yupResolver } from '@hookform/resolvers';
export { useForm } from 'react-hook-form';
export {
  handleArrayToObject,
  handleYupShape,
  handleYupExtractor,
  handleFilterFalsy,
} from './tools';
