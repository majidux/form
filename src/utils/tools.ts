import { yup } from './';

/** This function doesn't care about your given data type as long as its an array */
type ArrayToObjectType<GenericType> = GenericType & unknown[];
/** ArrayToObject function takes an array and turns it to a key-value pair of an object until it hits the end of array */
export function handleArrayToObject<GenericType>(
  data: ArrayToObjectType<GenericType>
): unknown {
  return data?.reduce((output: any, current: any) => ({
    ...output,
    ...current,
  }));
}

type YupShapeType = any;
/** handleYupShape function takes an object which was created before ArrayToObjectType */
export function handleYupShape(data: YupShapeType): any {
  return yup.object().shape({ ...data });
}

type ExtractorType<GenericType> = GenericType & {
  list: GenericType[];
  basedOn: keyof GenericType;
  keyName: keyof GenericType;
};

export function handleYupExtractor<Type>({
  list,
  basedOn,
  keyName,
}: ExtractorType<Type>): unknown[] {
  return list?.map(
    (item: Type) =>
      item[basedOn] && {
        [item[keyName] as any]: yup.string().required(item[basedOn] as any),
      }
  );
}

type FilterFalsyType = { list: unknown[] };
/** handleFilterFalsy function takes an array and delete all falsy parameters and returns a new array */
export function handleFilterFalsy({ list }: FilterFalsyType): unknown[] {
  return list?.filter(Boolean);
}
