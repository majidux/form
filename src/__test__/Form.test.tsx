import { Form } from '../utils';
import '@testing-library/jest-dom/extend-expect';
import { render, cleanup } from '@testing-library/react';

afterEach(cleanup);

test('renders correctly', () => {
  const onSubmit = (data: any) => {
    console.log(data);
  };
  const { getByTestId, asFragment } = render(<Form {...{ onSubmit, list }} />);
  expect(getByTestId('form')).toHaveTextContent('Send');
  expect(asFragment(<Form {...{ onSubmit, list }} />)).toMatchSnapshot();
});

const list = [
  {
    name: 'userName',
    require: 'Please enter your User name',
  },
  {
    value: 'Send',
    type: 'submit',
    itemType: 'button',
  },
];
