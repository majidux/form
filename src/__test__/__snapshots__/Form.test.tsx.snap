// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<DocumentFragment>
  <form
    class="form_parent"
    data-testid="form"
  >
    <section
      class="form_element-wrapper d-flex dir-column "
    >
      <input
        name="userName"
        placeholder="placeholder..."
        type="text"
      />
    </section>
    <button
      class="form-button"
    >
      Send
    </button>
  </form>
</DocumentFragment>
`;
